File: sources/common/src/main/java/fr/unice/polytech/si5/al/e/model/Travel.java
package fr.unice.polytech.si5.al.e.model;

import org.json.JSONArray;
import org.json.JSONObject;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.util.HashSet;
import java.util.Set;

@Entity
public class Travel {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private int id;

    @NotNull
    private String departure;

    @NotNull
    private String destination;

    private String state;

    @OneToOne(cascade = CascadeType.ALL)
    private Validator validator;

    @OneToMany(cascade = CascadeType.PERSIST, fetch = FetchType.EAGER)
    private Set<Item> items;

    @ManyToOne(cascade = CascadeType.ALL)
    private Customer customer;

    @ManyToOne(cascade = CascadeType.ALL)
    private Customer transporter;

    public Travel() {

        validator = new Validator();
        this.items = new HashSet<>();
    }

    public void addItem(Item item) {
        items.add(item);
    }

    public Set<Item> getItems() {
        return items;
    }

    public Customer getCustomer() {
        return customer;
    }

    public void setCustomer(Customer customer) {
        this.customer = customer;
    }

    public Customer getTransporter() {
        return transporter;
    }

    public void setTransporter(Customer transporter) {
        this.transporter = transporter;
    }

    public String getDeparture() {
        return departure;
    }

    public void setDeparture(String departure) {
        this.departure = departure;
    }

    public String getDestination() {
        return destination;
    }

    public void setDestination(String destination) {
        this.destination = destination;
    }

    public String getState() {
        return state;
    }

    public void setState(String state) {
        this.state = state;
    }

    public Validator getValidator() {
        return validator;
    }

    public int getId() {
        return id;
    }

    public String toJSON() {
        JSONObject object = new JSONObject();
        object.put("id", id);
        object.put("departure", departure);
        object.put("destination", destination);
        if (!items.isEmpty()) {
            JSONArray itemsJson = new JSONArray();
            items.forEach(i -> itemsJson.put(i.getName()));
            object.put("items", itemsJson);
        }
        object.put("customer", customer.getName());
        if (transporter != null) {
            object.put("transporter", transporter.getName());
        }
        object.put("state", state);

        return object.toString();

    }
}

File: blablamoveService/typescript/main/src/entity/Travel.ts
import { Comparable } from "./Comparable";
import { Item } from "./Item";
import { Customer } from "./Customer";
import { Validator } from "./Validator";

export class Travel implements Comparable {
    private id: string;
    private departure: string;
    private destination: string;
    private validator: Validator;
    private items: Item[];
    private customer: Customer;
    private transporter: Customer;

    constructor() {
        this.validator = new Validator();
        this.items = [];
    }

    public get $id(): string {
        return this.id;
    }
    public set $id(value: string) {
        this.id = value;
    }

    public get $departure(): string {
        return this.departure;
    }

    public set $departure(value: string) {
        this.departure = value;
    }

    public get $destination(): string {
        return this.destination;
    }

    public set $destination(value: string) {
        this.destination = value;
    }

    public get $items(): Item[] {
        return this.items;
    }

    public set $items(value: Item[]) {
        this.items = value;
    }
    public get $customer(): Customer {
        return this.customer;
    }

    public set $customer(value: Customer) {
        this.customer = value;
    }

    public get $transporter(): Customer {
        return this.transporter;
    }

    public set $transporter(value: Customer) {
        this.transporter = value;
    }

    public get $validator(): Validator {
        return this.validator;
    }

    public set $validator(value: Validator) {
        this.validator = value;
    }

    public addItem(item: Item) {
        this.items.push(item);
    }

    equal(object: any): boolean {
        if (!(object instanceof Travel)) return false;
        return this.id === object.id;
    }
}
